### ISSUE TITLE
Mixed precision training and LayerNormalization

### ISSUE THREAD

**Issue by gcuder** [tensorflow/models#10381]
Hello everyone,

is it safe to use `LayerNormalization` with `mixed_precision` training? I am referring to [this comment].(https://github.com/tensorflow/models/blob/bd7732dac80fbb2c6c62212e75e578d34bea760c/official/nlp/modeling/layers/transformer_encoder_block.py#L154)

**Comment by frederick0329** [comment]
Tagging mixed expert, Reed, here.

**Comment by frederick0329** [comment]
Discussed with Reed offline, we think it's fine but haven't have a chance to validate this. Feel free to try it on your use case and report back.

**Comment by gcuder** [comment]
I am sorry that it took me that long, but I tried using `mixed_precision` and I had a `NaN` loss during training. I attached the `tf.debugging.enable_check_numerics` debugger which yielded the following result

```

!!! Detected Infinity or NaN in output 0 of graph op "ReadVariableOp" (# of outputs: 1) !!!
  dtype: <dtype: 'float16'>
  shape: (512,)

  Input tensor: Tensor("evolved_transformer/evolved_transformer_encoder/layer_0/glu_layer_norm/batchnorm/mul/ReadVariableOp/resource:0", shape=(), dtype=resource)
  Graph name: "train_function"

  Stack trace of op's creation ("->": inferred user code):
    + ... (Omitted 14 frames)
    + ...xt2text/core/modeling/layers/transformer_blocks.py (L500) call
 -> |   if self._norm_first:
    + ...sorflow/python/autograph/operators/control_flow.py (L1174) if_stmt
    |   _py_if_stmt(cond, body, orelse)
    + ...sorflow/python/autograph/operators/control_flow.py (L1227) _py_if_stmt
    |   return body() if cond else orelse()
    + ...xt2text/core/modeling/layers/transformer_blocks.py (L501) call
 -> |   input_tensor = self._glu_layer_norm(input_tensor)
    + ...python3.8/site-packages/keras/engine/base_layer.py (L1037) __call__
 -> |   outputs = call_fn(inputs, *args, **kwargs)
    + .../keras/layers/normalization/layer_normalization.py (L288) call
 -> |   outputs = tf.nn.batch_normalization(
    + ...8/site-packages/tensorflow/python/util/dispatch.py (L206) wrapper
    |   return target(*args, **kwargs)
    + ...3.8/site-packages/tensorflow/python/ops/nn_impl.py (L1568) batch_normalization
    |   inv *= scale
    + ....8/site-packages/tensorflow/python/ops/math_ops.py (L1367) binary_op_wrapper
    |   return func(x, y, name=name)
    + ....8/site-packages/tensorflow/python/ops/math_ops.py (L1710) _mul_dispatch
    |   return multiply(x, y, name=name)
    + ...8/site-packages/tensorflow/python/util/dispatch.py (L206) wrapper
    |   return target(*args, **kwargs)
    + ....8/site-packages/tensorflow/python/ops/math_ops.py (L530) multiply
    |   return gen_math_ops.mul(x, y, name)
    + ...ite-packages/tensorflow/python/ops/gen_math_ops.py (L6245) mul
    |   _, _, _op, _outputs = _op_def_library._apply_op_helper(
    + ...ages/tensorflow/python/framework/op_def_library.py (L517) _apply_op_helper
    |   values = ops.convert_to_tensor(
    + .../site-packages/tensorflow/python/profiler/trace.py (L163) wrapped
    |   return func(*args, **kwargs)
    + ...8/site-packages/tensorflow/python/framework/ops.py (L1566) convert_to_tensor
    |   ret = conversion_func(value, dtype=dtype, name=name, as_ref=as_ref)
    + ...te-packages/tensorflow/python/distribute/values.py (L1376) _tensor_conversion_distributed_var
    |   return var._dense_var_to_tensor(dtype=dtype, name=name, as_ref=as_ref)  # pylint: disable=protected-access
    + ...te-packages/tensorflow/python/distribute/values.py (L1174) _dense_var_to_tensor
    |   return ops.convert_to_tensor(
    + .../site-packages/tensorflow/python/profiler/trace.py (L163) wrapped
    |   return func(*args, **kwargs)
    + ...8/site-packages/tensorflow/python/framework/ops.py (L1566) convert_to_tensor
    |   ret = conversion_func(value, dtype=dtype, name=name, as_ref=as_ref)
    + ...ges/tensorflow/python/ops/resource_variable_ops.py (L2023) _dense_var_to_tensor
    |   return var._dense_var_to_tensor(dtype=dtype, name=name, as_ref=as_ref)  # pylint: disable=protected-access
    + ...ges/tensorflow/python/ops/resource_variable_ops.py (L1421) _dense_var_to_tensor
    |   return self.value()
    + ...ges/tensorflow/python/ops/resource_variable_ops.py (L576) value
    |   return self._read_variable_op()
    + ...ges/tensorflow/python/ops/resource_variable_ops.py (L683) _read_variable_op
    |   result = read_and_set_handle()
    + ...ges/tensorflow/python/ops/resource_variable_ops.py (L673) read_and_set_handle
    |   result = gen_resource_variable_ops.read_variable_op(
    + ...tensorflow/python/ops/gen_resource_variable_ops.py (L484) read_variable_op
    |   _, _, _op, _outputs = _op_def_library._apply_op_helper(
    + ...ages/tensorflow/python/framework/op_def_library.py (L748) _apply_op_helper
    |   op = g._create_op_internal(op_type_name, inputs, dtypes=None,
    + ...packages/tensorflow/python/framework/func_graph.py (L599) _create_op_internal
    |   return super(FuncGraph, self)._create_op_internal(  # pylint: disable=protected-access
    + ...8/site-packages/tensorflow/python/framework/ops.py (L3561) _create_op_internal
    |   ret = Operation(
    + ...8/site-packages/tensorflow/python/framework/ops.py (L2045) __init__
    |   self._traceback = tf_stack.extract_stack_for_node(self._c_op)

 : Tensor had NaN values
	 [[node evolved_transformer/evolved_transformer_encoder/layer_0/glu_layer_norm/batchnorm/mul/ReadVariableOp/CheckNumericsV2 (defined at /tmp/pycharm_project_213/text2text/core/modeling/layers/transformer_blocks.py:501) ]] [Op:__inference_train_function_46214]
```
